File: Project_Cab_Fare_Prediction.R
# Cab Fare Prediction

setwd('D:/Data Science Edwisor/6. Projects/1. Project One Cab Rental Company')
# loading the dataset.
dataset = read.csv('train_cab.csv', na.strings = c(" ","", "NA"))

str(dataset)

# Extracting Meaningful Information from pickup_datetime.
library(lubridate)
dataset$pickup_datetime = ymd_hms(dataset$pickup_datetime)
dataset$pickup_date = day(dataset$pickup_datetime)
dataset$pickup_day = wday(dataset$pickup_datetime, label = FALSE)
dataset$pickup_month = month(dataset$pickup_datetime)
dataset$pickup_hour = hour(dataset$pickup_datetime)
str(dataset)

# Missing Value Analysis
sum(is.na(dataset))
apply(dataset, 2, function(x) {sum(is.na(x))})
# dropping the missing values
library(DataCombine)
dataset = DropNA(dataset)

# Outlier Analysis
# 1. Inspecting the latitudes and longitudes.
dataset[dataset['pickup_latitude'] < -90,]
dataset[dataset['pickup_latitude'] > 90,]
# Droping the row no 5687
dataset[dataset['pickup_latitude'] > 90,] = NA
dataset = DropNA(dataset)
dataset[dataset['dropoff_latitude'] < -90,]
dataset[dataset['dropoff_latitude'] > 90,]
dataset[dataset['pickup_longitude'] < -180,]
dataset[dataset['pickup_longitude'] > 180,]
dataset[dataset['dropoff_longitude'] < -180,]
dataset[dataset['dropoff_longitude'] > 180,]

# 2. fare_amount
dataset$fare_amount = as.numeric(as.character(dataset$fare_amount)) #Conversion generates one NA for 430- value in fare_amount.
summary(dataset$fare_amount)
dataset[order(-dataset$fare_amount),]
# Removing values > 150
dataset = DropNA(dataset)
dataset[dataset['fare_amount'] > 150,] = NA
dataset = DropNA(dataset)
dataset[order(dataset$fare_amount),'fare_amount']
# Removing values < 0
dataset[dataset['fare_amount'] < 0,] = NA
dataset = DropNA(dataset)
summary(dataset$fare_amount)
# Removing fare_amount = 0
dataset[dataset['fare_amount'] == 0,] = NA
dataset = DropNA(dataset)

# 3. Passenger_count
summary(dataset$passenger_count)
dataset[order(-dataset$passenger_count),'passenger_count']
# Removing passenger_count > 6
dataset[dataset['passenger_count'] > 6,] = NA
dataset = DropNA(dataset)
# Removing Passenger_count = 0
dataset[dataset['passenger_count'] == 0,] = NA
dataset = DropNA(dataset)
# Removing Passenger_count = 0.12
dataset[dataset['passenger_count'] == 0.12,] = NA
dataset = DropNA(dataset)
# Removing Passenger_count = 1.3
dataset[dataset['passenger_count'] == 1.3,] = NA
dataset = DropNA(dataset)

# Defining Haversine Distance Function
distance = function(lat1, lon1, lat2, lon2){
      p = 0.017453292519943295 # Pi/180
      a = 0.5 - cos((lat2 - lat1) * p)/2 + cos(lat1 * p) * cos(lat2 * p) * (1 - cos((lon2 - lon1) * p)) / 2
      d = 12742 * sin(sqrt(a))
      return(d)
}

dataset$distance_travelled = distance(dataset$pickup_latitude, dataset$pickup_longitude,
                                      dataset$dropoff_latitude, dataset$dropoff_longitude)
# Outlier Analysis for distance_travelled
dataset[order(-dataset$distance_travelled),'distance_travelled']
# removing some very high value in distance_travelled. i.e. distance_travelled > 1000
dataset[dataset['distance_travelled'] > 1000,] = NA
dataset = DropNA(dataset)
sum(is.na(dataset$distance_travelled))
summary(dataset$distance_travelled)
# Now max distance is 129.946
nrow(dataset[dataset['distance_travelled'] == 0,])
nrow(dataset[dataset['fare_amount'] == 0,])
nrow(dataset[(dataset['distance_travelled'] == 0) & (dataset['fare_amount'] > 2.5),])
nrow(dataset[(dataset['distance_travelled'] == 0) & (dataset['fare_amount'] == 2.5),])
nrow(dataset[(dataset['distance_travelled'] == 0) & (dataset['fare_amount'] < 2.5),])
# removing distance_travelled = 0 and fare amount > 2.5(assuming cancallation charge is 2.5)
dataset[(dataset['distance_travelled'] == 0) & (dataset['fare_amount'] > 2.5),'distance_travelled'] = NA
sum(is.na(dataset$distance_travelled))
dataset = DropNA(dataset)

# Removing pickup_datetime, and variables with longitudes and latitudes.
dataset = subset(dataset, select = -c(pickup_datetime, pickup_longitude, pickup_latitude, 
                                     dropoff_longitude, dropoff_latitude))
# Visualisations
# Scatter Plot visualisation for different variables.
library(ggplot2)

# 1. Passenger_count
ggplot() + 
  geom_point(aes(x = dataset$passenger_count, y = dataset$fare_amount),
             colour = 'red') +
  ggtitle(' Fare Vs Passenger_Count') +
  xlab('No of Passengers') +
  ylab('Fare')

# 2. Pickup_Date
ggplot() + 
  geom_point(aes(x = dataset$pickup_date, y = dataset$fare_amount),
             colour = 'red') +
  ggtitle(' Fare Vs Pickup_Date') +
  xlab('Date of Travel') +
  ylab('Fare')

# 3. Pickup_day
ggplot() + 
  geom_point(aes(x = dataset$pickup_day, y = dataset$fare_amount),
             colour = 'red') +
  ggtitle(' Fare Vs Pickup_Day') +
  xlab('Day of Travel') +
  ylab('Fare')

# 4. Pickup_month
ggplot() + 
  geom_point(aes(x = dataset$pickup_month, y = dataset$fare_amount),
             colour = 'red') +
  ggtitle(' Fare Vs Pickup_Month') +
  xlab('Month of Travel') +
  ylab('Fare')

# 5. Pickup_Hour
ggplot() + 
  geom_point(aes(x = dataset$pickup_hour, y = dataset$fare_amount),
             colour = 'red') +
  ggtitle(' Fare Vs Pickup_Hour') +
  xlab('Hour of Travel') +
  ylab('Fare')

# 6. Distance_Travelled
ggplot() + 
  geom_point(aes(x = dataset$distance_travelled, y = dataset$fare_amount),
             colour = 'red') +
  ggtitle(' Fare Vs Distance_Travelled') +
  xlab('Distacne Travelled') +
  ylab('Fare')

# Checking Distribution
# 1. fare_Amount
plot(density(dataset$fare_amount))
# we find that our data is skewed.
# Applying log to remove skewness.
dataset$fare_amount = log1p(dataset$fare_amount)
# Recheck
plot(density(dataset$fare_amount))

# 2. passenger_Count
plot(density(dataset$passenger_count))

# 3. pickup_date
plot(density(dataset$pickup_date))

# 4. pickup_day
plot(density(dataset$pickup_day))

# 5. pickup_month
plot(density(dataset$pickup_month))

# 6. pickup_hour
plot(density(dataset$pickup_hour))

# 7. distance_travelled
plot(density(dataset$distance_travelled))
# We see skewness in this data
# Removing skewness
dataset$distance_travelled = log1p(dataset$distance_travelled)
# Recheck
plot(density(dataset$distance_travelled))

# Converting the data into proper datatypes
dataset$passenger_count = as.factor(dataset$passenger_count)
dataset$pickup_day = as.factor(dataset$pickup_day)
dataset$pickup_hour = as.factor(dataset$pickup_hour)

# Splitting the data into train and test
library(caTools)

set.seed(123)
split = sample.split(dataset$fare_amount, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Building Models

# 1. Multiple Linear Regression Model
regressor_LR = lm(formula = fare_amount ~ .,
               data = training_set)
summary(regressor_LR)

# Predicting the test set
y_pred = predict(regressor_LR, newdata = test_set)

# MAPE CHECk
MAPE = function(y, x){
  mean(abs((y - x)/y)) * 100
}

MAPE(test_set[, 1], y_pred)

# Alternate method
library(DMwR)
regr.eval(test_set[,1], y_pred, stats = c('rmse', 'mape'))

# 2. Decision Tree Model
library(rpart)
regressor_DT = rpart(formula = fare_amount ~ .,
                     data = training_set,
                     control = rpart.control(minsplit = 7))
summary(regressor_DT)

# Predicting the test set
y_pred = predict(regressor_DT, newdata = test_set)

# Evaluating Performance
regr.eval(test_set[,1], y_pred, stats = c('rmse', 'mape'))

# 3. Random Forest Model
# Fitting Random Forest Regression Model to the model
library(randomForest)
set.seed(1234)
regressor_RF = randomForest(x = training_set[, 2:7],
                         y = training_set$fare_amount,
                         ntree = 500)
# Predicting the test set
y_pred = predict(regressor_RF, newdata = test_set)
# Evaluating Performance
regr.eval(test_set[,1], y_pred, stats = c('rmse', 'mape'))

# Finalising Random Forest model for our Cab Fare Prediction based on better rmse value
# and mape value is also very close to that of linear regression model.

# Reading and preparing the test set
dataset_test = read.csv('test.csv', na.strings = c(" ","", "NA"))

# Taking care of pickup_datetime variable.
dataset_test$pickup_datetime = ymd_hms(dataset_test$pickup_datetime)
dataset_test$pickup_date = day(dataset_test$pickup_datetime)
dataset_test$pickup_day = wday(dataset_test$pickup_datetime, label = FALSE)
dataset_test$pickup_month = month(dataset_test$pickup_datetime)
dataset_test$pickup_hour = hour(dataset_test$pickup_datetime)
str(dataset_test)

# Missing Value Analysis
sum(is.na(dataset_test))

# Outlier Analysis
# 1. Inspecting the latitudes and longitudes.
dataset_test[dataset_test['pickup_latitude'] < -90,]
dataset_test[dataset_test['pickup_latitude'] > 90,]
dataset_test[dataset_test['dropoff_latitude'] < -90,]
dataset_test[dataset_test['dropoff_latitude'] > 90,]
dataset_test[dataset_test['pickup_longitude'] < -180,]
dataset_test[dataset_test['pickup_longitude'] > 180,]
dataset_test[dataset_test['dropoff_longitude'] < -180,]
dataset_test[dataset_test['dropoff_longitude'] > 180,]

# 2. Passenger_count
summary(dataset_test$passenger_count)
dataset_test[order(-dataset_test$passenger_count),'passenger_count']
# None found

# Defining Haversine Distance Function
distance = function(lat1, lon1, lat2, lon2){
  p = 0.017453292519943295 # Pi/180
  a = 0.5 - cos((lat2 - lat1) * p)/2 + cos(lat1 * p) * cos(lat2 * p) * (1 - cos((lon2 - lon1) * p)) / 2
  d = 12742 * sin(sqrt(a))
  return(d)
}

dataset_test$distance_travelled = distance(dataset_test$pickup_latitude, dataset_test$pickup_longitude,
                                      dataset_test$dropoff_latitude, dataset_test$dropoff_longitude)
# Outlier Analysis for distance_travelled
dataset_test[order(-dataset_test$distance_travelled),'distance_travelled']
summary(dataset_test$distance_travelled)

# Removing pickup_datetime, and variables with longitudes and latitudes.
dataset_test = subset(dataset_test, select = -c(pickup_datetime, pickup_longitude, pickup_latitude, 
                                      dropoff_longitude, dropoff_latitude))
# Normality Check
# 1. distance_travelled
plot(density(dataset_test$distance_travelled))
# We see skewness in this data
# Removing skewness
dataset_test$distance_travelled = log1p(dataset_test$distance_travelled)
# Recheck
plot(density(dataset_test$distance_travelled))
# # 2. passenger_Count
# plot(density(dataset_test$passenger_count))
# # 3. pickup_date
# plot(density(dataset_test$pickup_date))
# # 4. pickup_day
# plot(density(dataset_test$pickup_day))
# # 5. pickup_month
# plot(density(dataset_test$pickup_month))
# # 6. pickup_hour
# plot(density(dataset_test$pickup_hour))
# Except distance_travelled none other variables are skewed

# Converting the data into proper datatypes
dataset_test$passenger_count = as.factor(dataset_test$passenger_count)
dataset_test$pickup_day = as.factor(dataset_test$pickup_day)
dataset_test$pickup_hour = as.factor(dataset_test$pickup_hour)

# 1. Final Random Forest Model
# Training Random Forest Regression Model to the whole training set
library(randomForest)
set.seed(1234)
regressor_RF_final = randomForest(x = dataset[, 2:7],
                            y = dataset$fare_amount,
                            ntree = 500)
y_pred_rf = predict(regressor_RF_final, newdata = dataset)
# Evaluating Performance
regr.eval(dataset[,1], y_pred_rf, stats = c('rmse', 'mape'))

# Making Final Predictions on our test set
y_pred_final = predict(regressor_RF_final, newdata = dataset_test)

# # Storing back the output
# cab_fare_prediction = data.frame(Predicted_Fares = y_pred_final)
# write.csv(cab_fare_prediction, 'cab fare predictions R.csv')

############################### END #####################################################

File: Project_Cab_Fare_Prediction.py
# Importing the liabraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os
import seaborn as sns
#% matplotlib inline 

#setting working directory
os.chdir("D:/Data Science Edwisor/6. Projects/1. Project One Cab Rental Company")

# Importing the dataset
dataset = pd.read_csv('train_cab.csv')

# Exploratory Data Analysis
# Converting to proper data type
dataset['fare_amount'] = pd.to_numeric(dataset['fare_amount'], errors = 'coerce')
dataset['pickup_datetime'] = pd.to_datetime(dataset['pickup_datetime'], errors = 'coerce')
dataset['passenger_count'] = dataset['passenger_count'].astype('object')
# Extracting useful information from 'pickup_datetime'
dataset['pickup_date'] = dataset['pickup_datetime'].dt.day
dataset['pickup_day'] = dataset['pickup_datetime'].dt.weekday
dataset['pickup_month'] = dataset['pickup_datetime'].dt.month
dataset['pickup_hour'] = dataset['pickup_datetime'].dt.hour
dataset['pickup_minute'] = dataset['pickup_datetime'].dt.minute
dataset['pickup_day'] = dataset['pickup_day'].astype('object')
dataset['pickup_hour'] = dataset['pickup_hour'].replace(0, 24)

# Removing pickup_datetime variable from dataset
dataset = dataset.iloc[:, [0,2,3,4,5,6,7,8,9,10,11]]

# Inspecting the Latitudes and longitudes
# Latitude(-90 - 90)
dataset[dataset['pickup_latitude'] < -90]
dataset[dataset['pickup_latitude'] > 90]
dataset = dataset.drop(dataset[dataset['pickup_latitude'] > 90].index, axis = 0)

dataset[dataset['dropoff_latitude'] < -90]
dataset[dataset['dropoff_latitude'] > 90]

# Longitude(-180 - 180)
dataset[dataset['pickup_longitude'] < -180]
dataset[dataset['pickup_longitude'] > 180]

dataset[dataset['dropoff_longitude'] < -180]
dataset[dataset['dropoff_longitude'] > 180]


# Data Preprocessing

# 1. Missing Value Analysis.
dataset.isnull().sum()

"""# Finding the best method for missing value imputation.
dataset['pickup_longitude'].loc[7]
# Actual Value is -73.9513
dataset['pickup_longitude'].loc[7] = np.nan

# Mean
dataset['pickup_longitude'] = dataset['pickup_longitude'].fillna(dataset['pickup_longitude'].mean())
# Mean value = -73.912

# Median
dataset['pickup_longitude'] = dataset['pickup_longitude'].fillna(dataset['pickup_longitude'].median())
# Median value = -73.982

# KNN
from fancyimpute import KNN
imputer = KNN(k = 5)
dataset = pd.DataFrame(imputer.fit_transform(dataset), columns = dataset.columns)
# KNN Value = -73.993
# Imputing with median
from sklearn.preprocessing import Imputer
imputer = Imputer(missing_values = 'NaN', strategy = 'median', axis = 0)
imputer = imputer.fit(dataset.iloc[:, [0,1,2,3,4,6,8,9,10]])
dataset.iloc[:, [0,1,2,3,4,6,8,9,10]] = imputer.transform(dataset.iloc[:, [0,1,2,3,4,6,8,9,10]])
# Imputing with mode
imputer_mode = Imputer(missing_values = 'NaN', strategy = 'most_frequent', axis = 0)
imputer_mode = imputer_mode.fit(dataset.iloc[:, [5,7]])
dataset.iloc[:, [5,7]] = imputer_mode.transform(dataset.iloc[:, [5,7]])
dataset.isnull().sum()
dataset['passenger_count'] = dataset['passenger_count'].astype('object')
dataset['pickup_day'] = dataset['pickup_day'].astype('object')"""
# Dropping the values with NAN. 
dataset = dataset.drop(dataset[dataset['pickup_date'].isnull()].index, axis = 0)
dataset = dataset.drop(dataset[dataset['fare_amount'].isnull()].index, axis = 0)
dataset = dataset.drop(dataset[dataset['passenger_count'].isnull()].index, axis = 0)
dataset.isnull().sum()
# 2. Outiler Analysis.

# 1. Fare_amount
dataset['fare_amount'].describe()
dataset['fare_amount'].sort_values(ascending = False)
# Some very high values present in fare_amount
# Removing values > 150
dataset = dataset.drop(dataset[dataset['fare_amount'] > 150].index, axis = 0)
# removing values < 0
dataset = dataset.drop(dataset[dataset['fare_amount'] < 0].index, axis = 0)
dataset['fare_amount'].sort_values(ascending = True)
# Removing fare_amount = 0
dataset = dataset.drop(dataset[dataset['fare_amount'] == 0].index, axis = 0)


# 2. Passenger_count
dataset['passenger_count'].describe()
dataset['passenger_count'].sort_values(ascending = False)
# Some Very High values in this variable as well
# removing values > 6
dataset = dataset.drop(dataset[dataset['passenger_count'] > 6].index, axis = 0)
# removing passenger_count = 0
dataset = dataset.drop(dataset[dataset['passenger_count'] == 0].index, axis = 0)
dataset['passenger_count'].sort_values(ascending = True)
# removing passenger_count = 0.12
dataset = dataset.drop(dataset[dataset['passenger_count'] == 0.12].index, axis = 0)

# Extracting Meaningful information(Distance) from coordinates.
def distance(lat1, lon1, lat2, lon2):
    p = 0.017453292519943295 # Pi/180
    a = 0.5 - np.cos((lat2 - lat1) * p)/2 + np.cos(lat1 * p) * np.cos(lat2 * p) * (1 - np.cos((lon2 - lon1) * p)) / 2
    return 12742 * np.arcsin(np.sqrt(a))

dataset['distance_travelled'] = distance(dataset['pickup_latitude'], dataset['pickup_longitude'],
                          dataset['dropoff_latitude'], dataset['dropoff_longitude'])
# Outlier Analysis For Ditance_travelled
dataset['distance_travelled'].sort_values(ascending = False)
dataset['distance_travelled'].describe()
# Removing values more than 1000kms
dataset = dataset.drop(dataset[dataset['distance_travelled'] > 1000].index, axis = 0)
dataset['distance_travelled'].describe()
# Now max distance is 129.95
from collections import Counter 
Counter(dataset['distance_travelled'] == 0)
Counter(dataset['fare_amount'] == 0)
Counter((dataset['distance_travelled'] == 0) & (dataset['fare_amount'] > 2.5))
Counter((dataset['distance_travelled'] == 0) & (dataset['fare_amount'] == 2.5))
# Removing distance_travelled = 0 from dataset where fare is > 2.5
dataset = dataset.drop(dataset[(dataset['distance_travelled'] == 0) & (dataset['fare_amount'] > 2.5)].index, axis = 0)

# Removing the pickup_longitude, pickup_latitude, dropoff_longitude, dropoff_latitude from dataset
dataset = dataset.iloc[:, [0,5,6,7,8,9,10,11]]


# Feature Selection
# Correlation Analysis(between numeric variables)
subset = dataset.iloc[:, [2,4,5,6,7]]

f, ax = plt.subplots(figsize = (7, 5))
corr = subset.corr()
sns.heatmap(corr, mask = np.zeros_like(corr, dtype = np.bool), 
            cmap = sns.diverging_palette(220, 10, as_cmap = True),
           square = True, ax = ax)


# Visualisation
# Scatter Plot visualisation for different variables

# 1. Passenger_Count
plt.figure(figsize=(14,7))
plt.scatter(x = dataset['passenger_count'], y = dataset['fare_amount'], s=10)
plt.xlabel('No. of Passengers')
plt.ylabel('Fare')
plt.show()
# 2. Pickup_date
plt.figure(figsize=(14,7))
plt.scatter(x = dataset['pickup_date'], y = dataset['fare_amount'], s=10)
plt.xlabel('Date')
plt.ylabel('Fare')
plt.show()
# 3. Pickup_day
plt.figure(figsize=(14,7))
plt.scatter(x = dataset['pickup_day'], y = dataset['fare_amount'], s=10)
plt.xlabel('Pickup Day')
plt.ylabel('Fare')
plt.show()
# 4. Pickup_month
plt.figure(figsize=(14,7))
plt.scatter(x = dataset['pickup_month'], y = dataset['fare_amount'], s=10)
plt.xlabel('Pickup Month')
plt.ylabel('Fare')
plt.show()
# 5. Pickup_hour
plt.figure(figsize=(15,7))
dataset.groupby(dataset["pickup_hour"])['pickup_hour'].count().plot(kind="bar")
plt.show()

plt.figure(figsize=(14,7))
plt.scatter(x = dataset['pickup_hour'], y = dataset['fare_amount'], s=10)
plt.xlabel('Pickup Hour')
plt.ylabel('Fare')
plt.show()
# 6. Pickup_minute
plt.figure(figsize=(14,7))
plt.scatter(x = dataset['pickup_minute'], y = dataset['fare_amount'], s=10)
plt.xlabel('Pickup Minute')
plt.ylabel('Fare')
plt.show()
# 7. Distance_travelled
plt.figure(figsize=(14,7))
plt.scatter(x = dataset['distance_travelled'], y = dataset['fare_amount'], s=10)
plt.xlabel('Distance Travelled')
plt.ylabel('Fare')
plt.show()

# Normality Check
for i in ['fare_amount']:
    print(i)
    sns.distplot(dataset[i],bins='auto',color='blue')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()  
for i in ['distance_travelled']:
    print(i)
    sns.distplot(dataset[i],bins='auto',color='blue')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()
for i in ['passenger_count']:
    print(i)
    sns.distplot(dataset[i],bins='auto',color='blue')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()
for i in ['pickup_date']:
    print(i)
    sns.distplot(dataset[i],bins='auto',color='blue')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()
for i in ['pickup_day']:
    print(i)
    sns.distplot(dataset[i],bins='auto',color='blue')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()
for i in ['pickup_month']:
    print(i)
    sns.distplot(dataset[i],bins='auto',color='blue')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()
for i in ['pickup_hour']:
    print(i)
    sns.distplot(dataset[i],bins='auto',color='blue')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()
for i in ['pickup_minute']:
    print(i)
    sns.distplot(dataset[i],bins='auto',color='blue')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()
    
# Removing Skewness in distance_travelled and fare_amount
dataset['distance_travelled'] = np.log1p(dataset['distance_travelled'])
dataset['fare_amount'] = np.log1p(dataset['fare_amount'])
# Recheck for Normality
for i in ['fare_amount']:
    print(i)
    sns.distplot(dataset[i],bins='auto',color='blue')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()
for i in ['distance_travelled']:
    print(i)
    sns.distplot(dataset[i],bins='auto',color='blue')
    plt.title("Distribution for Variable "+i)
    plt.ylabel("Density")
    plt.show()
    
# Seperating independent and dependent variables.
X = dataset.iloc[:, 1:].values
Y = dataset.iloc[:, 0].values

# SPLITTING THE DATA INTO TRAIN AND TEST.
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

# Building Models

# 1. Multiple Linear Regression Model
from sklearn.linear_model import LinearRegression
regressor_LR = LinearRegression()
regressor_LR.fit(X_train, Y_train)
# Predicting on Test Set
Y_pred = regressor_LR.predict(X_test)

# Calculating Mape
def MAPE(true, pred):
    mape = np.mean(np.abs((true - pred)/true))* 100
    return mape

MAPE(Y_test, Y_pred)

# RMSE
from sklearn import metrics
print(np.sqrt(metrics.mean_squared_error(Y_test, Y_pred)))

# R Square

# Train Data
Y_pred_TR = regressor_LR.predict(X_train) 
from sklearn.metrics import r2_score
r2_score(Y_train, Y_pred_TR)

# Test Data
r2_score(Y_test, Y_pred)

# 2. Decision Tree 
# Fitting Decision Tree Regression Model
from sklearn.tree import DecisionTreeRegressor
regressor_DT = DecisionTreeRegressor(max_depth = 10, random_state = 0)
regressor_DT.fit(X_train, Y_train)

# Predicting on Test Set
Y_pred = regressor_DT.predict(X_test)

# Calculating Mape
MAPE(Y_test, Y_pred)

# RMSE
print(np.sqrt(metrics.mean_squared_error(Y_test, Y_pred)))

# R Square
# Train Data
Y_pred_TR = regressor_DT.predict(X_train) 
r2_score(Y_train, Y_pred_TR)

# Test Data
r2_score(Y_test, Y_pred)

# 3. Random Forest
from sklearn.ensemble import RandomForestRegressor
regressor_RF = RandomForestRegressor(max_depth = 7, n_estimators = 300, random_state = 1)
regressor_RF.fit(X_train, Y_train)

# Predicting on Test Set
Y_pred = regressor_RF.predict(X_test)

# Calculating Mape
MAPE(Y_test, Y_pred)

# RMSE
print(np.sqrt(metrics.mean_squared_error(Y_test, Y_pred)))

# R Square
# Train Data
Y_pred_TR = regressor_RF.predict(X_train) 
r2_score(Y_train, Y_pred_TR)

# Test Data
r2_score(Y_test, Y_pred)

# As Random Forest has the best values for RMSE and Rsquared hence 
# lets make it as the final model.
# Let's try some parameter tuning for RF.

# Applying K-Fold Cross Validation for Random Forest
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = regressor_RF,
                             X = X_train,
                             y = Y_train,
                             cv = 7)
# calculating the mean of obtained accuracies.
accuracies.mean()

"""# Grid Search
from sklearn.model_selection import GridSearchCV
parameters = [{'max_depth' : [5,7,9], 'n_estimators' : [300, 400, 500],
               'random_state' : [0,1,2]}]
grid_search = GridSearchCV(estimator = regressor_RF,
                           param_grid = parameters,
                           cv = 7,
                           n_jobs = -1)
grid_search = grid_search.fit(X, Y)

# getting the scores and parameters
best_accuracy = grid_search.best_score_
best_parameters = grid_search.best_params_"""
# Commenting the above code as it takes time to run.

# The grid search suggests n_estimators = 300, random_state = 1. Lets try this on our
# regressor_RF and recheck the results.

# Finalising RandomForest Model for cab fare prediction on our actual Test
# Set.

# Reading and preparing the test set data for our model

dataset_test = pd.read_csv('test.csv')
# Converting to proper data types
dataset_test['pickup_datetime'] = pd.to_datetime(dataset_test['pickup_datetime'], errors = 'coerce')
dataset_test['passenger_count'] = dataset_test['passenger_count'].astype('object')
# Extracting useful information from 'pickup_datetime'
dataset_test['pickup_date'] = dataset_test['pickup_datetime'].dt.day
dataset_test['pickup_day'] = dataset_test['pickup_datetime'].dt.weekday
dataset_test['pickup_month'] = dataset_test['pickup_datetime'].dt.month
dataset_test['pickup_hour'] = dataset_test['pickup_datetime'].dt.hour
dataset_test['pickup_minute'] = dataset_test['pickup_datetime'].dt.minute
dataset_test['pickup_day'] = dataset_test['pickup_day'].astype('object')
dataset_test['pickup_hour'] = dataset_test['pickup_hour'].replace(0, 24)

# Removing pickup_datetime variable from dataset
dataset_test = dataset_test.iloc[:, [1,2,3,4,5,6,7,8,9,10]]

# Inspecting the Latitudes and longitudes
# Latitude(-90 - 90)
dataset_test[dataset_test['pickup_latitude'] < -90]
dataset_test[dataset_test['pickup_latitude'] > 90]

dataset_test[dataset_test['dropoff_latitude'] < -90]
dataset_test[dataset_test['dropoff_latitude'] > 90]

# Longitude(-180 - 180)
dataset_test[dataset_test['pickup_longitude'] < -180]
dataset_test[dataset_test['pickup_longitude'] > 180]

dataset_test[dataset_test['dropoff_longitude'] < -180]
dataset_test[dataset_test['dropoff_longitude'] > 180]
# None are beyond the range.

# 1. Missing Value Analysis.
dataset_test.isnull().sum()
# None Found

# 2. Exploring the variables
# 1. Passenger_count
dataset_test['passenger_count'].describe()
dataset_test['passenger_count'].sort_values(ascending = False)
dataset_test['passenger_count'].sort_values(ascending = True)

# Extracting Meaningful information(Distance) from coordinates.
def distance(lat1, lon1, lat2, lon2):
    p = 0.017453292519943295 # Pi/180
    a = 0.5 - np.cos((lat2 - lat1) * p)/2 + np.cos(lat1 * p) * np.cos(lat2 * p) * (1 - np.cos((lon2 - lon1) * p)) / 2
    return 12742 * np.arcsin(np.sqrt(a))

dataset_test['distance_travelled'] = distance(dataset_test['pickup_latitude'], dataset_test['pickup_longitude'],
                          dataset_test['dropoff_latitude'], dataset_test['dropoff_longitude'])
# Outlier Analysis For Ditance_travelled
dataset_test['distance_travelled'].sort_values(ascending = False)
dataset_test['distance_travelled'].describe()
# None found

# Removing the pickup_longitude, pickup_latitude, dropoff_longitude, dropoff_latitude from dataset
dataset_test = dataset_test.iloc[:, 4:11]

# Visualisations
# Normality Check
for i in ['distance_travelled']:
    print(i)
    sns.distplot(dataset_test[i],bins='auto',color='blue')
    plt.title("Distribution for Test Dataset Variable "+i)
    plt.ylabel("Density")
    plt.show()
# Removing Skewness in distance_travelled
dataset_test['distance_travelled'] = np.log1p(dataset_test['distance_travelled'])
# Normality ReCheck
for i in ['distance_travelled']:
    print(i)
    sns.distplot(dataset_test[i],bins='auto',color='blue')
    plt.title("Distribution for Test Dataset Variable "+i)
    plt.ylabel("Density")
    plt.show()

# Now building the final model with whole training data
from sklearn.ensemble import RandomForestRegressor
regressor_RF_final = RandomForestRegressor(max_depth = 7, n_estimators = 300, random_state = 1)
regressor_RF_final.fit(X, Y)

Y_pred_final = regressor_RF_final.predict(X) 
# Calculating Mape(Training Data)
MAPE(Y, Y_pred_final)
# MAPE obtained 8.41%, Accuracy = 92%

# RMSE(Training Data)
print(np.sqrt(metrics.mean_squared_error(Y, Y_pred_final)))

# R Square(Training Data)
r2_score(Y, Y_pred_final)

# Predicting on Test Set
predicted_fares = regressor_RF_final.predict(dataset_test)

"""# storing the predicted fares.
Predicted_fares = pd.DataFrame({'Predicted_Fares': predicted_fares})
Predicted_fares.to_csv('Final Predicted Fares.csv', index = False)"""

# Deleting temp variables
del i
del X_test
del X_train
del Y_pred
del Y_pred_TR
del Y_test
del Y_train
del accuracies
del subset

############################# END #######################################